core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategy.java

```diff
@@ -0,0 +1,140 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint.healthcheck;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+
+import org.jctools.maps.NonBlockingHashSet;
+
+import com.linecorp.armeria.client.Endpoint;
+
+import com.google.common.collect.ImmutableSet;
+
+public class PartialHealthCheckStrategy implements HealthCheckStrategy {
+
+    private final Set<Endpoint> selectedCandidates;
+
+    private final double percentage;
+    private final int minimumSelectedCandidates;
+
+    private List<Endpoint> candidates;
+
+    private volatile boolean change;
+
+    public PartialHealthCheckStrategy(double percentage, int minimumSelectedCandidates) {
+        checkArgument(percentage > 0 && percentage <= 100, "percentage: %s (expected: 0.x-100)", percentage);
+        this.percentage = percentage;
+        checkArgument(minimumSelectedCandidates > 0, "minimumSelectedCandidates: %s (expected: 1 - INT_MAX)",
+                      minimumSelectedCandidates);
+        this.minimumSelectedCandidates = minimumSelectedCandidates;
+        selectedCandidates = new NonBlockingHashSet<>();
+        candidates = new ArrayList<>();
+        change = false;
+    }
+
+    @Override
+    public void updateCandidate(List<Endpoint> candidates) {
+        synchronized (selectedCandidates) {
+            this.candidates = requireNonNull(candidates, "candidates");
+
+            final Set<Endpoint> removedCandidates = new HashSet<>();
+            for (Endpoint candidate : this.candidates) {
+                if (selectedCandidates.contains(candidate)) {
+                    continue;
+                }
+
+                removedCandidates.add(candidate);
+            }
+
+            updateSelectedCandidates(removedCandidates);
+        }
+    }
+
+    @Override
+    public List<Endpoint> getCandidates() {
+        if (!change) {
+            return NOT_CHANGE;
+        }
+
+        synchronized (selectedCandidates) {
+            change = false;
+
+            return new ArrayList<>(selectedCandidates);
+        }
+    }
+
+    @Override
+    public void updateHealth(Endpoint endpoint, double health) {
+        final double unhealthyScore = 0;
+
+        requireNonNull(endpoint, "endpoint");
+
+        if (health > unhealthyScore) {
+            return;
+        }
```

Comment:
Do we need to add the endpoint back to `selectedCandidates` here if we do not have enough number of available endpoints at the moment?

Reply:
I think almost the candidates' health is healthy.

So if we change PartialHealthCheckStrategy's status(selectedCandidates) when it's healthy, we use too many locks.

To avoid that, I designed that update when only it's unhealthy.